### ISSUE TITLE
Feature request: mass rename

### ISSUE THREAD

**Issue by calexicoz** [Jermolene/TiddlyWiki5#2946]
Hi,

I'm a fairly recent user of TiddlyWiki, and I use the (latest) node version. I have adopted a naming scheme for my tiddlers that follows typical filesystem conventions. Ie, "root" tiddlers have a regular name, and "children" tiddlers are named "parent/child". So for example, I would have:

```
projects.tid
projects/foo.tid
projects/bar.tid
projects/bar/note.tid
```

If I rename `projects`, I'll get the prompt to automatically rename all the tags and lists (Update OLD to NEW in the tags and list fields of other tiddlers). This won't however change the **title** of all children (and grandchildren, ..). So if I rename `projects` to `my-projects` for example, I'll end up with this:

```
my-projects.tid
projects/foo.tid
projects/bar.tid
projects/bar/note.tid
```

The tags will be correct, and the TOC will still behave as expected, but what I would like instead is this:

```
my-projects.tid
my-projects/foo.tid
my-projects/bar.tid
my-projects/bar/note.tid
```

Basically I just want to follow filesystem conventions. I don't know how many people use such a naming scheme, it's not really the "wiki way" (if there is such a thing).

How would I go about automatically make such changes all the time? Seems like I could create a plugin with a `$tw.hooks.addHook("th-renaming-tiddler", function(newTiddler,oldTiddler)`, but I want to make sure that's the right way to go about it.

Thanks!

**Comment by pmario** [comment]
>How would I go about automatically make such changes all the time? Seems like I could create a plugin with a $tw.hooks.addHook("th-renaming-tiddler", function(newTiddler,oldTiddler), but I want to make sure that's the right way to go about it.

If you know, how to do this, it's the right approach.  .... but .... 

----

I personally wouldn't do the whole renaming thing at all. For me the advantages are not worth the problems it causes.

TW file names are basically independent from the tiddler title. Exactly for the reason you describe. Keeping names in sync is hard. As you found out. 

TW core uses a slightly different approach. The file system groups tiddlers into logical groups. Internally they are grouped by the namespace. .. but the namespace doesn't really correspond to the file system structure. This removes the need, to keep it in sync.

**Comment by sukima** [comment]
Many filesystems I use won't allow the tiddler title as the filename. And I wouldn't want it to.

For example I have posts in my TiddlyWiki in files like `tiddlers/posts/2017-07-21-short-description.tid` While the actual title is like:

```
title: Very Witty and Descriptive Title
tags: Article Coding

…
```

I would not want my title tied to my filesystem.

**Comment by sukima** [comment]
Also if you create the file first you can refrence it by setting thr `retain-original-tiddler-path` config to `true`. See [tiddlywiki.info Files](http://tiddlywiki.com/#tiddlywiki.info%20Files), maybe that is what you're looking for?

**Comment by calexicoz** [comment]
Hi all,

Thanks for the replies. I feel like I need to expand a little bit on my actual problem, of which the filesystem is only a symptom. In the end, it all comes down to this: how to organize a large amount of content in a structured manner.

In the following, I'm talking about titles (while I actually want to keep my filesystem structured the same way, this is not the core issue). So let's go through a couple examples.


## Example 1

I keep a journal of my daily work, so I'll start with something like this:

```
journal/2017-07-21
journal/2017-07-22
journal/2017-07-23
```

After a while, I'll want to also do a personal journal, so I'll want to restructure things like this:

```
perso/journal/2017-07-21
perso/journal/2017-07-22
perso/journal/2017-07-23
work/journal/2017-07-21
work/journal/2017-07-22
work/journal/2017-07-23
```

It could be argued that I could leave the original work journal as is, and just add the new content:
```
perso/journal/2017-07-21
perso/journal/2017-07-22
perso/journal/2017-07-23
journal/2017-07-21
journal/2017-07-22
journal/2017-07-23
```

While it certainly possible, I find it an unsuitable choice because it is now not clear what type of journal `journal/XXX` is. And yes I wrote it, and I'll probably know and remember that it's the work one, but it just rubs me the wrong way.


## Example 2

So one of the main tenant of TiddlyWiki (and other wiki systems) is that titles are unique. So I can't name 2 things "notes". So if I want to keep notes about my projects, I'll naturally break them up this way:

```
project1/notes
project2/notes
```

It happens that some of my projects are used by multiple customers, and I need to keep track of different things for each of them, so at some point I'll have to break things up further:

```
customerA/project1/notes
customerA/project2/notes
customerB/project1/notes
customerB/project2/notes
```

At this point "notes" appears 4 times, and projectX twice, so they both need some kind of extra identifier to make their title unique. Following filesystem conventions of just using a "hierarchy/path" and a `/` separator is just one natural way to keep titles unique, and the content generally structured.

Meanwhile, there might actually be content at each level, so that the TOC looks like this:

```
customerA
	/project1
		/notes
	/project2
		/notes
customerB
	/project1
		/notes
	/project2
		/notes
```

## Main point

I guess the original issue simply brings up those questions:

- How to handle structuring content at scale
- How to handle refactoring of content (including structure)

I think TiddlyWiki is unique in the Wiki world because it encourages "smaller" units for content, and refactoring of content is definitely a part of that. But the smaller the units, the larger the number of items, and the greater the need for organization.

Certainly, some people do well without structure, and are happy to have all their title be simple and straightforward, at the root, with no hierarchy, and will rely on search to find their eggs. I'm sadly not one of those people, and structure is important to me. In particular, it helps when I go to "all pages", because I see this:

```
large-category1/concept1/...
large-category1/concept1/...
large-category1/concept1/...
large-category1/concept2/...
large-category1/concept2/...

large-category2/project1...
large-category2/project1...
large-category2/project1...

large-category3/itemA...
large-category3/itemB...
```

Because things are structured, I can know/see/understand at a glance what is actually in my wiki. What are the big pieces. How do they fit together. If I simply used descriptive titles (which could go around the unicity constraint), I would end up with this:

```
How to renew credentials
```

Instead of this:

```
Coding/AWS/Snippets/Identity/How-to-renew-credentials
```

The 2nd is a lot more helpful to me because I know that in the long run I'll have:

```
Coding/AWS/Snippets/EC2/...
Coding/AWS/Snippets/ELBs/...
Coding/AWS/Snippets/Identity/...
Coding/AWS/Snippets/S3/...
```

Often times, when looking for a snippet, I won't remember what I named it exactly, but will know it exists, and so will search for `snippets/s3` for example as a quick way to get to a filtered-down list that I can go through instantly.

Any way, this is a little long-winded, but I hope it brings some clarity to my actual concern. Advice welcome :)

**Comment by calexicoz** [comment]
Also, in that last example, `Coding/AWS/Snippets/S3/...` is useful itself, because `AWS`, `snippets` and `S3` will show up in a lot of other tiddlers over time, and only the combinations of them will make sense to me when I want to find something. But I may not get to that structure initially, because I may only have 2 or 3 snippets to start with. This structure will arise from the need to organize a larger and larger number of small items, that may all be otherwise incredibly similar with only minor variations.

**Comment by pmario** [comment]
OK ... I do see your structure ... and structuring by namespace is a valid approach. .... but imo TW isn't optimized for it, if you want to "mirror" that structure to the file system. It causes a lot of syncing problems. .. As you found out. 

My personal approach for your examples would be with `tags` and `fields`. ... Those elements are first class citizens in TW. 

eg: you can create a tiddler named: `projects` which contains something similar to this: 

```
<<list-links "[tag[projects]]">>
```

Then you can use a tags eg: `project1, project2` ... both tagged: `projects` ... So the above code will dynamically create a list of projects. ... 

`project1` can also be tagged: `customer-1` ... I personally wouldn't name 2 different projects `project1`. For the same reason, why I wouldn't name 2 different customers with `customer-1`.  Technically, there is no difference here. ... With customers, it's just more obvious.

customer-1 tiddler could contain: 

```
<<list-links "[tag[customer-1]]">>
```

`journal/2017-07-22` ... can be tagged with any tags of the above. IMO there is no need for a namespace. If I want fine granularity, I would use hh:mm if needed. 

--------

If you don't want to use tags, you could use fields instead. Sorting / filtering possibilities would be similar. ... 

Just some ideas.

**Comment by Jermolene** [comment]
Hi @calexicoz have you seen http://tiddlywiki.com/#Customising%20Tiddler%20File%20Naming? The mechanism allows you to specify rules for generating the filenames for new (or renamed) tiddlers, and may help you get closer to what you want.

**Comment by calexicoz** [comment]
@pmario I'll look into that approach, though I doubt it will work for me. I think that like tabs vs spaces, or vim vs emacs, tags (or other types of fields) vs folders is definitely a battle of two camps, and two ideologies, of which I fall on the folder side. Both are definitely valid ways to organize data, but my mental model is just happier with hierarchies of things.

@Jermolene I did look into this, and I'm using this feature already. Unfortunately, that doesn't quite solve the requirement for unicity of names for tiddlers. I could otherwise have modified the file naming scheme, and even the search result tiddlers to show me "generated names", like "parent / parent / tiddler". But I'll still end up with two tiddlers named "notes", or "conceptX".

My best bet at the moment is to write a simple python/ruby script to do mass renaming for me by directly manipulating the tiddler files, while the server is off. While refactors do happen as content grows, they are occasional, so that might be "Good Enough (TM)".

**Comment by calexicoz** [comment]
Thanks all for the help. Looks like I'll have to try a few different things. Closing this now. Cheers!
